### PR TITLE
introduce a Transport

### LINKED ISSUE #3797: race condition in key update test
https://app.circleci.com/pipelines/github/quic-go/quic-go/5968/workflows/5c3ef88b-f755-415d-8e1e-849f05195240/jobs/12200

### PR DESCRIPTION
**Pull request by marten-seemann**
Part of #3727. Fixes #3797.

### CODE REVIEW

**Review by lucas-clemente — APPROVED** [review]
> add a GetConfigForClient callback to the Config ### TODO: add tests

one of the commits :)

File: transport.go [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+
+		if t.ConnectionIDGenerator != nil {
+			t.connIDLen = t.ConnectionIDGenerator.ConnectionIDLen()
+		} else {
+			connIDLen := t.ConnectionIDLength
+			if t.ConnectionIDLength == 0 && (!t.isSingleUse || isServer) {
+				connIDLen = protocol.DefaultConnectionIDLength
+			}
+			t.connIDLen = connIDLen
+			t.ConnectionIDGenerator = &protocol.DefaultConnectionIDGenerator{ConnLen: t.connIDLen}
```
Comment (ydnar) [inline transport.go:225]: Perhaps this should be copied to a private `connectionIDGenerator`?

Doesn’t feel right to mutate a public field which could be changed later, which would have no effect on the `Transport`.

File: transport.go [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+	// If the connection satisfies the OOBCapablePacketConn interface
+	// (as a net.UDPConn does), ECN and packet info support will be enabled.
+	// In this case, optimized syscalls might be used, skipping the
+	// ReadFrom and WriteTo calls to read / write packets.
+	Conn net.PacketConn
+
+	// The length of the connection ID in bytes.
+	// It can be 0, or any value between 4 and 18.
+	// If unset, a 4 byte connection ID will be used.
+	ConnectionIDLength int
```
Comment (ydnar) [inline transport.go:37]: Maybe omit `ConnectionIDLength` entirely in favor of `ConnectionIDGenerator`?

File: transport.go [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+		t.conn = conn
+		t.handlerMap = newPacketHandlerMap(t.StatelessResetKey, t.enqueueClosePacket, t.logger)
+		t.listening = make(chan struct{})
+
+		t.closeQueue = make(chan closePacket, 4)
+
+		if t.ConnectionIDGenerator != nil {
+			t.connIDLen = t.ConnectionIDGenerator.ConnectionIDLen()
+		} else {
+			connIDLen := t.ConnectionIDLength
```
Comment (ydnar) [inline transport.go:220]: If `(Transport).ConnectionIDLength` was removed, this branch on `t.isSingleUse` and `isServer` can be eliminated.

File: transport.go (hunk 1 above)
Comment (marten-seemann) [reply transport.go:225]: That fields shouldn't be modified later applies to all fields, but I see your point here. The application might check if it set the field, and this won't work any more. I'll make the change.

File: packet_handler_map.go [hunk 4]
```diff
@@ -5,28 +5,22 @@ import (
...
 	"sync"
 	"time"
 
 	"github.com/quic-go/quic-go/internal/protocol"
 	"github.com/quic-go/quic-go/internal/utils"
-	"github.com/quic-go/quic-go/internal/wire"
-	"github.com/quic-go/quic-go/logging"
 )
 
-// rawConn is a connection that allow reading of a receivedPacket.
```
[the remark this answers is not in the document]
Comment (marten-seemann) [reply packet_handler_map.go:25]: Typo in a removed comment?

File: transport.go (hunk 2 above)
Comment (marten-seemann) [reply transport.go:37]: I consider `ConnectionIDGenerator` somewhat of a fringe use case. You only need it if you want to use a Connection ID-based load balancer.

File: transport.go [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+		case <-t.listening:
+			return
+		case p := <-t.closeQueue:
+			t.conn.WritePacket(p.payload, p.addr, p.info.OOB())
+		}
+	}
+}
+
+// Close closes the underlying connection and waits until listen has returned.
+func (t *Transport) Close() error {
```
[the remark this answers is not in the document]
Comment (marten-seemann) [reply transport.go:255]: Nope. I'll improve documentation.

**Comment by marten-seemann** [comment]
> > add a GetConfigForClient callback to the Config ### TODO: add tests
> 
> one of the commits :)

Added tests. Which caught several bugs :)

**Pull request merged**